chia/wallet/vc_wallet/vc_wallet.py

```diff
@@ -0,0 +1,460 @@
+        if chia_tx is not None and chia_tx.spend_bundle is not None:
+            spend_bundles.append(chia_tx.spend_bundle)
+            tx_list.append(dataclasses.replace(chia_tx, spend_bundle=None))
+        spend_bundle = SpendBundle.aggregate(spend_bundles)
+        now = uint64(int(time.time()))
+        add_list: List[Coin] = list(spend_bundle.additions())
+        rem_list: List[Coin] = list(spend_bundle.removals())
+        tx_list.append(
+            TransactionRecord(
+                confirmed_at_height=uint32(0),
+                created_at_time=now,
+                to_puzzle_hash=new_inner_puzhash,
+                amount=uint64(1),
+                fee_amount=uint64(fee),
+                confirmed=False,
+                sent=uint32(0),
+                spend_bundle=spend_bundle,
+                additions=add_list,
+                removals=rem_list,
+                wallet_id=self.id(),
+                sent_to=[],
+                trade_id=None,
+                type=uint32(TransactionType.OUTGOING_TX.value),
+                name=spend_bundle.name(),
+                memos=list(compute_memos(spend_bundle).items()),
+            )
+        )
+        return tx_list
+
+    async def revoke_vc(
+        self, parent_id: bytes32, peer: WSChiaConnection, fee: uint64 = uint64(0), reuse_puzhash: Optional[bool] = None
+    ) -> List[TransactionRecord]:
+        vc_coin_states: List[CoinState] = await self.wallet_state_manager.wallet_node.get_coin_state(
+            [parent_id], peer=peer
+        )
+        if vc_coin_states is None:
+            raise ValueError(f"Cannot find verified credential coin: {parent_id.hex()}")
+        vc_coin_state = vc_coin_states[0]
+        cs: CoinSpend = await fetch_coin_spend_for_coin_state(vc_coin_state, peer)
+        vc: VerifiedCredential = VerifiedCredential.get_next_from_coin_spend(cs)
+
+        # Check if we own the DID
+        found_did = False
+        for _, did_wallet in self.wallet_state_manager.wallets.items():
+            if did_wallet.type() == WalletType.DECENTRALIZED_ID:
+                assert isinstance(did_wallet, DIDWallet)
+                if bytes32.fromhex(did_wallet.get_my_DID()) == vc.proof_provider:
+                    found_did = True
+                    break
+        if not found_did:
+            raise ValueError(f"You don't own the DID {vc.proof_provider.hex()}")
+
+        _, provider_inner_puzhash, _ = await did_wallet.get_info_for_recovery()
+
+        # Generate spend specific nonce
+        coins = await did_wallet.select_coins(uint64(1))
+        assert coins is not None
+        coins.add(vc.coin)
+        if fee > 0:
+            coins.update(await self.standard_wallet.select_coins(fee))
+        sorted_coins: List[Coin] = sorted(coins, key=Coin.name)
+        sorted_coin_list: List[List[Union[bytes32, uint64]]] = [coin_as_list(c) for c in sorted_coins]
+        nonce: bytes32 = Program.to(sorted_coin_list).get_tree_hash()
```

Comment:
why are we generating nonce via coin selection? also, why do we run separate select coin for a fee? given we're not ignoring previously selected coins, won't we get the same coins back potentially?

Reply:
> why are we generating nonce via coin selection
This is an anti replay measure.  We want the announcement that we assert for bundle cohesion to have a nonce specific to the spend we are creating so that if another overlapping spend is created they cannot both be pushed just due to a single announcement.  Rather, only one spend can go through because they are both depending on a unique announcement and the VC can only make one of those two.

> also, why do we run separate select coin for a fee
These coins are just being selected for nonce calculation, so they're not being used in any way.  That being said, this should be the same exact call that `create_tandem_xch_tx` makes internally and will thus select the coins that we used to calculate the nonce for use.